contracts/data/DoublyLinkedList.sol

```diff
@@ -0,0 +1,428 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title Doubly linked list implementation with enumeration functions
+ */
+library DoublyLinkedList {
+    struct DoublyLinkedListInternal {
+        mapping(bytes32 => bytes32) _nextValues;
+        mapping(bytes32 => bytes32) _prevValues;
+    }
```

Comment:
As per our earlier discussion; it may be worth considering adding a `head` and `tail` pointer to the struct which point to the first and last items of the DLL, respectively. 

For Example:

```md
head <-> A <-> B <-> C <-> D <-> tail
```

In this case `head == A`, `tail == D`. If `D` is removed from the DLL, `tail == C`  


```md
head <-> A <-> B <-> C <-> tail
```

Reply:
In order for us to be able to query first/last element, we would have to store head/tail in a separate variable.  This means an additional storage write on every `push` (which I believe will be the most commonly used operation).